src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -0,0 +1,219 @@
+      if (lower_bound_value_expression != nullptr && upper_bound_value_expression != nullptr) {
+        const auto between_node = PredicateNode::make(
+            std::make_shared<BetweenExpression>(current_column_expression, lower_bound_value_expression,
+                                                upper_bound_value_expression, left_inclusive, right_inclusive));
+        between_nodes.push_back(between_node);
+      } else {
+        predicate_nodes.insert(std::end(predicate_nodes), std::begin(node_scope), std::end(node_scope));
+      }
+      upper_bound_value_expression = nullptr;
+      lower_bound_value_expression = nullptr;
+      left_inclusive = true;
+      right_inclusive = true;
+      node_scope = std::vector<std::shared_ptr<AbstractLQPNode>>();
+      current_column_expression = boundary.column_expression;
+    }
+
+    if (boundary.type == ColumnBoundaryType::UpperBoundaryInclusive) {
+      if (upper_bound_value_expression == nullptr ||
+          upper_bound_value_expression->value > boundary.value_expression->value) {
+        upper_bound_value_expression = boundary.value_expression;
+        right_inclusive = true;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::LowerBoundaryInclusive) {
+      if (lower_bound_value_expression == nullptr ||
+          lower_bound_value_expression->value < boundary.value_expression->value) {
+        lower_bound_value_expression = boundary.value_expression;
+        left_inclusive = true;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::UpperBoundaryExclusive) {
+      if (upper_bound_value_expression == nullptr ||
+          upper_bound_value_expression->value <= boundary.value_expression->value) {
+        upper_bound_value_expression = boundary.value_expression;
+        right_inclusive = false;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::LowerBoundaryExclusive) {
+      if (lower_bound_value_expression == nullptr ||
+          lower_bound_value_expression->value <= boundary.value_expression->value) {
+        lower_bound_value_expression = boundary.value_expression;
+        left_inclusive = false;
+        node_scope.push_back(boundary.node);
+      }
+    }
+  }
+
+  // Apply boundary check again for the last column
+  if (lower_bound_value_expression != nullptr && upper_bound_value_expression != nullptr) {
+    const auto between_node = PredicateNode::make(
+        std::make_shared<BetweenExpression>(current_column_expression, lower_bound_value_expression,
+                                            upper_bound_value_expression, left_inclusive, right_inclusive));
+    between_nodes.push_back(between_node);
+  } else {
+    predicate_nodes.insert(std::end(predicate_nodes), std::cbegin(node_scope), std::cend(node_scope));
```

Comment:
Consistency: We use `x.begin()` instead of `std::begin(x)`, even if the latter is not wrong.